src/persist/src/operators/stream.rs

```diff
@@ -269,16 +270,17 @@ where
                 });
 
                 if !active_seal_operator {
-                    return;
+                    continue;
```

Comment:
initial question while I'm digesting the actual implementation bits: if this is expected to always be used in a `while scheduler.yield_now().await` loop, can we push that loop into build_async (or offer a variant of build_async that does it)? it seems somewhat idiomatic in rust for macros to hide early returns (e.g. `try!/?`, `bail!`, `ready!`) and if we can, we may as well lean into that instead of having to be careful about only using `continue`s

Reply:
I've been grappling with this too, I don't like the `continue` here at all. There is one ugly solution of doing this:

```rust
while scheduler.yield_now().await {
    async {
        if early_return_condition {
            return;
        }
    }.await;
}
```

Which allows us to use `return`, `bail!` etc but increases even more the boilerplate. Hiding these details inside `build_async` I think is going to make the API horrible for the following reasoning:

1. If `build_async` itself builds the future containing the while loop, then it must be given a logic function that can produce the inside future on demand like so: ` async { while ... { logic().await } }`
2. Since that logic function must be callable in a loop it cannot be an `FnOnce`, it must be an `FnMut` or an `Fn`
3. The useful futures that we want to return from this logic function will almost certainly be capturing variables from that closure's environment. This is not allowed and the compiler will complain that closure state is escaping. Then the only solution for the API user could be to use some kind of `Rc<RefCell<..>>` for every single captured closure state, and clone the, every single time they produce a future.

I'll keep ruminating on this in case a better idea comes along

Reply:
Here is the error I'm talking about in isolation: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d3774452f48e02cc95dc55f1771bdb5c

Reply:
Yeah that's compelling. If we stick with the while loop, I actually prefer the additional boilerplate. I'll take "clunky but easy to reason about" over "more elegant but more subtle" any day

This is a shot in the dark, but would it work to use a 2-level closure to solve the Fn ownership issues?

Reply:
> This is a shot in the dark, but would it work to use a 2-level closure to solve the Fn ownership issues?

Ah I see, maybe not. This works for timely operators because the logic closure gets called in a loop, but those communicate by emitting things not by returning them, so we still have a (3) issue

(Whoops, wrote this in the wrong place the first time.)

Reply:
> I actually prefer the additional boilerplate. I'll take "clunky but easy to reason about" over "more elegant but more subtle" any day

Definitely agree! I hadn't thought of the extra boilerplate solution until your comment so thanks for the nudge :) I'll change the PR to that style

Btw, the real solution is to not rely on closures to provide the logic but rather define an `AsyncOperatorLogic` trait that captures this lifetime relationship and provide an `L: AsyncOperatorLogic` struct to build_async. Making timely async-friendly is a goal  I'm making slow progress on so we might see that in the future